cwltool/workflow_job.py

```diff
@@ -879,3 +883,223 @@ def job(
             self.do_output_callback(output_callback)
             # depends which one comes first. All steps are completed
             # or all outputs have been produced.
+
+
+class WorkflowJobLoopStep:
+    """Generated for each step in Workflow.steps() containing a http://commonwl.org/cwltool#Loop requirement."""
+
+    def __init__(self, step: WorkflowJobStep, container_engine: str):
+        """Initialize this WorkflowJobLoopStep."""
+        self.step: WorkflowJobStep = step
+        self.container_engine: str = container_engine
+        self.joborder: Optional[CWLObjectType] = None
+        self.processStatus: str = "success"
+        self.iteration: int = 0
+        self.output_buffer: MutableMapping[
+            str,
+            Union[MutableSequence[Optional[CWLOutputType]], Optional[CWLOutputType]],
+        ] = {}
+
+    def job(
+        self,
+        joborder: CWLObjectType,
+        output_callback: OutputCallbackType,
+        runtimeContext: RuntimeContext,
+    ) -> JobsGeneratorType:
+        """Generate a WorkflowJobStep job until the `loopWhen` condition evaluates to False."""
+        self.joborder = joborder
+        loop_req = cast(
+            CWLObjectType,
+            self.step.step.get_requirement("http://commonwl.org/cwltool#Loop")[0],
+        )
+
+        callback = functools.partial(
+            self.loop_callback,
+            runtimeContext,
+        )
+
+        try:
+            while True:
+                evalinputs = {shortname(k): v for k, v in self.joborder.items()}
+                whenval = expression.do_eval(
+                    loop_req["loopWhen"],
+                    evalinputs,
+                    self.step.step.requirements,
+                    None,
+                    None,
+                    {},
+                    debug=runtimeContext.debug,
+                    js_console=runtimeContext.js_console,
+                    timeout=runtimeContext.eval_timeout,
+                    container_engine=self.container_engine,
+                )
+                if whenval is True:
+                    yield from self.step.job(self.joborder, callback, runtimeContext)
+                    if self.processStatus == "permanentFail":
```

Comment:
When yield returns control to this method, you can't assume the step has actually finished running.  You only know that the step is finished when the callback has been called.

I suggest setting `processStatus = ""` before doing `yield from self.step.job(...)`

```
self.processStatus = ""
yield from self.step.job(...)
while self.processStatus == "":
  yield None
```

The loop callback needs to always set `self.processStatus = processStatus` to indicate the callback was called and the loop should proceed.

Reply:
This is not clear to me. Why do you say that the process could not be finished? There is no fire and forget logic inside the implementation.

Reply:
The job() method returns a python generator (iterator protocol).  It does one of three things.

1. yield a Job object (which has a run() method)
2. yield None (there's nothing to do right now, but there might be something to do in the future
3. return and end the iterator (the operation is completed)

The job object is invoked at some point in the future.  In particular, with the multi-threaded executor, it goes into a queue.

Meanwhile, we keep calling the iterator to get more work.  For example, a scatter step is going to produce N jobs, an then idle (return None) until all those jobs have completed.

At the top level of the workflow, we loop over each step and the ones which are ready to run, we call the job() method to ask that step to generate jobs for us.  If that step returns None we'll go on to the next thing, but we'll eventually loop around back to iterate on the step again.

It might be helpful to think of these iterators as channels feeding job objects up to the top level.  Part of the process is going through each iterator and asking it if it has new values to produce or not (sort of like reading a channel, except that we are actually passing control to the iterator which has to yield it back).  

Does this sort of make sense?